### PR TITLE
Custom dataset for SHL Creation

### PR DESCRIPTION
**Pull request by sarahmcdougall**
# Summary
Adds the ability for the user to create a SMART Health Link for a custom vaccine data set.

## New Behavior
The user can now create a new custom data set from all the available vaccines, and create a new sharing link for it. If the link is deleted, another one can be created since the data set will be saved to local storage.

## Code Changes
* New route “health-links/new-custom” for creating a custom dataset and then creating a SHLink for it
* Updates the reducer to add a new dataset 
* New helper function `filterForIds` to filter all the vaccines to those that should be included in a data set
* New component `SHCLinkCreateCustom` that displays all the vaccines, keeps track of check boxes for the vaccines, and allows for a new data set to be created along with a new SHLink for that dataset

# Testing Guidance
* Follow the [server setup instructions](https://github.com/dvci/vaxx.link/tree/main/server) from the vaxx.link/server directory. The server will run on port 8000.
* Run the UI with `npm start` from the vaxx.link/ui directory. The app will run on port 3000 and is accessible at http://localhost:3000/.
* Navigate to “Health Links” in the nav bar. There should be two default datasets already populated - All Vaccines and School Vaccines. There should be a third header that says “New Custom Data Set.” Click on the “Create new link” next to the “New Custom Data Set” header. 
* After clicking on the button, you should be navigated to a new page for selecting the vaccines to include in the new data set. Select some of the vaccines, then click “Activate New Sharing Link.”
* After clicking on the button, you will be brought back to the Health Links page. Check that a the custom data set has been created (its header defaults to Custom Data Set {data set id}.) A SHLink should be created for the data set. To check that the data set was appropriately filtered to just include the checked vaccines, try deactivating the SHLink and clicking on “Create new link” again. The options should be limited to those that were checked when creating the custom data set.

### CODE REVIEW

File: ui/src/App.tsx [hunk 1]
```diff
@@ -550,15 +598,18 @@ export function SHLinks() {
           ))}
         </React.Fragment>
       ))}
-
       <h4>
-        Custom Data Set
-        <button>Create new link</button>
+        New Custom Data Set
+        <button onClick={() => navigate('/health-links/new?custom=true', {state: {custom: true}})}>Create new dataset and link</button>
```
Comment (jmandel) [inline ui/src/App.tsx]: This line is duplicating the `custom: true` as a URL param and also in a "state" object.

Are there reasons not to just use `custom=true` as a URL param? Then you could do something similar to what we have in place for the dataset ID (`let datasetId = Number(searchParams.get('ds'));`) today.

File: ui/src/App.tsx [hunk 2]
```diff
@@ -550,15 +598,18 @@ export function SHLinks() {
...
       </h4>
     </div>
   );
 }
 
+interface LocationState {
```
Comment (jmandel) [inline ui/src/App.tsx]: See comment above about passing `custom=true` in url param

File: ui/src/App.tsx [hunk 3]
```diff
@@ -377,19 +385,45 @@ const defaultImmunizations: Promise<StoredSHC[]> = Promise.all(
...
 
   async function activate() {
+    if (custom) {
+      const checkedVaccinations = vaccines.map(card => card.id).filter((card, i) => isChecked[i] === true);
+      // create new DataSet using the checked vaccinations
+      // TODO: Allow user to rename the dataset
+      datasetId = Object.keys(store.sharing).length;
+      const customDataSet : DataSet = {
+        id: datasetId,
+        name: `Custom DataSet ${datasetId}`,
```
Comment (jmandel) [inline ui/src/App.tsx]: It'd be great to land this functionality before we merge -- may be as simple as just adding an `<input>` and populating its default value to `Custom Data Set ${datasetId}` but allowing the user to type their own value inside; then you could just grab that value when `activate()` is called.

File: ui/src/App.tsx [hunk 4]
```diff
@@ -377,19 +385,45 @@ const defaultImmunizations: Promise<StoredSHC[]> = Promise.all(
...
 export function SHLinkCreate() {
   let navigate = useNavigate();
-  let { store } = useStore();
+  let location = useLocation();
+  let { store, dispatch } = useStore();
   let [usePasscode, setUsePasscode] = useState(false);
   let [passcode, setPasscode] = useState('1234');
   let [expires, setExpires] = useState(false);
 
+  const { custom } = location.state as LocationState || {};
```
Comment (jmandel) [inline ui/src/App.tsx]:
```suggestion
  let custom = Boolean(searchParams.get('custom') === 'true');
```

**Review by jmandel — CHANGES_REQUESTED** [review]
Thanks so much for this submission! I have a couple of suggestions that I'd love to get in place before we merge, if they sound OK to you too.

File: ui/src/App.tsx (hunk 1 above)
Comment (sarahmcdougall) [reply ui/src/App.tsx]: I think we can just use `custom=true` as a URL param - I don't have much experience with URL params in React but after testing without the "state" object it still seems to work.

I pushed the rest of the changes as well, so this PR is ready for a re-review!

**Comment by jmandel** [comment]
Thanks so much for this contribution!

**Pull request merged**